shared/src/store/action-reducers/exercise.ts

```diff
@@ -171,3 +194,76 @@ function refreshTransfer(
         letElementArrive(draftState, type, element.id);
     });
 }
+
+export interface TreatmentAssignment {
+    [patientId: UUID]: { [personnelType in PersonnelType]: number };
+}
+
+function calculateTreatmentAssignment(
+    draftState: Mutable<ExerciseState>
+): TreatmentAssignment {
+    const treatmentAssignment = StrictObject.fromEntries(
+        Object.keys(draftState.patients).map((patientId) => [
+            patientId,
+            StrictObject.fromEntries(
+                StrictObject.keys(personnelTypeAllowedValues).map(
+                    (personnelType) => [personnelType, 0]
+                )
+            ),
+        ])
+    ) as TreatmentAssignment;
+
+    StrictObject.values(draftState.personnel).forEach((personnel) => {
+        const assignedPatientCount = StrictObject.keys(
+            personnel.assignedPatientIds
+        ).length;
+        StrictObject.keys(personnel.assignedPatientIds)
+            .filter((patientId) => treatmentAssignment[patientId])
+            .forEach((patientId) => {
+                treatmentAssignment[patientId]![personnel.personnelType]! +=
+                    1 / assignedPatientCount;
+            });
+    });
+
+    return treatmentAssignment;
+}
+
+function evaluateTreatmentReassignment(
+    draftState: Mutable<ExerciseState>,
+    newTreatmentAssignment: TreatmentAssignment
+) {
+    if (!draftState.previousTreatmentAssignment) return;
+
+    Object.keys(newTreatmentAssignment)
+        .filter((patientId) =>
+            StrictObject.keys(personnelTypeAllowedValues).some(
+                (personnelType) =>
+                    newTreatmentAssignment[patientId]![personnelType] !==
+                    draftState.previousTreatmentAssignment![patientId]?.[
+                        personnelType
+                    ]
+            )
+        )
+        .forEach((patientId) => {
+            logPatient(
+                draftState,
+                StrictObject.entries(newTreatmentAssignment[patientId]!)
+                    .filter(([, count]) => count > 0)
+                    .map(([personnelType]) =>
+                        createPersonnelTypeTag(draftState, personnelType)
+                    ),
+                `Die Einsatzkräfte zu diesem Patienten wurden neu zugeteilt. Er hat nun die folgenden Einsatzkräfte: ${StrictObject.entries(
+                    newTreatmentAssignment[patientId]!
+                )!
+                    .filter(([, count]) => count > 0)
+                    .map(
+                        ([personnelType, count]) =>
+                            `${count.toPrecision(2)} ${
+                                personnelTypeNames[personnelType]
+                            }`
+                    )
+                    .join(', ')}.`,
```

Comment:
I think you could omit the "er hat..." part, since it makes the message unnecessary long. Just say "... neu zugeteilt: X NA, Y NFS, ...".

You should also handle empty personnel (it currently just says "die folgenden Einsatzkräfte: .") and display integer values without a trailing ".0".